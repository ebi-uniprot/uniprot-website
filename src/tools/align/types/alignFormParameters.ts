import { ServerParameters } from './alignServerParameters';

export type FormParameters = {
  sequence: ServerParameters['sequence'];
};
