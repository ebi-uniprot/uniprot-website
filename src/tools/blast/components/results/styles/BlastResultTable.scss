@import '../../../../../../node_modules/franklin-sites/src/styles/colours';

@import '../../../../../shared/styles/settings';

.data-table {
  .query-sequence-wrapper {
    width: 32vw;
    margin-left: 0.2rem;
    // Hide unwanted items from navigation
    protvista-navigation {
      .zoom-polygon,
      .brush,
      .start-label,
      .end-label {
        display: none;
      }
    }
  }
  &__header-cell {
    top: 3rem !important;
  }
}

.data-table__blast-hsp {
  $self: &;
  display: flex;

  .chip {
    margin-top: 0;

    &.hsp_identity {
      --main-chip-color: #{$colour-sapphire-blue};
    }

    &.hsp_bit_score {
      --main-chip-color: #{$colour-coyote-brown};
    }

    &.hsp_expect {
      --main-chip-color: #{$colour-outer-space};
    }
  }

  &__tracks {
    position: relative;

    #{ $self }__blast-track {
      /**
      * Magic number preventing the forever growing
      * protvista-track
      */
      width: 32vw;
      cursor: pointer;

      &:hover {
        background: #d2dce3;
      }

      svg {
        margin: 0.2rem;
      }
    }

    #{ $self }__blast-params {
      position: absolute;
      top: -0.6rem;
      right: 0.1rem;
    }
  }

  button {
    margin-left: 0.5rem;
  }
}

@keyframes loading {
  from {
    opacity: 1;
  }
  to {
    opacity: 0.5;
  }
}

.loading-data-table .data-table {
  animation: alternate loading 1000ms infinite ease-in-out;
}
