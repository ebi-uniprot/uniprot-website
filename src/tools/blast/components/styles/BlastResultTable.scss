.data-table__blast-hsp {
  display: flex;
  button {
    white-space: nowrap;
  }
  &__tracks {
    width: 40vw;
    svg {
      margin: 0.1rem 0;
    }
  }
}

/**
* HACK! this should be fixed in franklin
*/
.data-table {
  input[type='checkbox'] {
    margin: 0;
  }
}
