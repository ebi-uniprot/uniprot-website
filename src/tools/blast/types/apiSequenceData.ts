type Name = {
  value: string;
};

type ProteinName = {
  fullName: Name;
  shortNames?: Name[];
};

export enum ProteinExistence {
  ProteinLevel = '1. Experimental evidence at protein level',
  TranscriptLevel = '2. Experimental evidence at transcript level',
  Inferred = '3. Protein inferred from homology',
  Predicted = '4. Protein predicted',
  Uncertain = '5. Protein uncertain',
}

export type APISequenceData = {
  entryType: string;
  primaryAccession: string;
  uniProtkbId: string;
  entryAudit: {
    sequenceVersion: number;
  };
  organism: {
    scientificName: string;
    commonName: string;
    taxonId: number;
    lineage: string[];
  };
  proteinExistence: ProteinExistence;
  proteinDescription: {
    recommendedName: ProteinName;
    alternativeNames: ProteinName[];
  };
  genes: {
    geneName: Name;
    synonyms: Name[];
  }[];
  sequence: {
    value: string;
    length: number;
    molWeight: number;
    /**
     * @deprecated Should use MD5 instead
     */
    crc64: string;
    md5: string;
  };
};
