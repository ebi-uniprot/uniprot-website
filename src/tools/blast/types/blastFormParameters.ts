import {
  SType,
  Database,
  Exp,
  Matrix,
  Filter,
  GapAlign,
  Scores,
  Alignments,
  Program,
  Sequence,
} from './blastServerParameters';
import { SelectedTaxon } from '../config/BlastFormData';

export type FormParameters = {
  stype: SType;
  program: Extract<Program, 'blastp' | 'blastx'>;
  sequence: Sequence;
  database: Database;
  taxIDs: SelectedTaxon[];
  negativeTaxIDs: SelectedTaxon[];
  threshold: Exp;
  matrix: Matrix;
  filter: Filter;
  gapped: GapAlign;
  hits: Scores | Alignments;
};
