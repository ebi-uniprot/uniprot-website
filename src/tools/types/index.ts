export enum Tool {
  blast = 'blast',
  // align = 'align',
  // peptideSearch = 'peptideSearch',
  // idMapping = 'idMapping',
}
