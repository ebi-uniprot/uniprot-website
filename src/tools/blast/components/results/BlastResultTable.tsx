/* eslint-disable @typescript-eslint/camelcase */
import React, { FC, Fragment, useCallback, useState } from 'react';
import { DataTable, DENSITY_COMPACT } from 'franklin-sites';
import { Link } from 'react-router-dom';
import ProtvistaTrack from 'protvista-track';
import { BlastResults, BlastHsp, BlastHit } from '../../types/blastResults';
import { loadWebComponent } from '../../../../shared/utils/utils';

import '../styles/BlastResultTable.scss';

const BlastSummaryTrack: FC<{
  hsp: BlastHsp;
  length: number;
}> = ({ hsp, length }) => {
  const { hsp_query_from, hsp_query_to, hsp_identity } = hsp;

  const setTrackData = useCallback(
    (node): void => {
      if (node) {
        // eslint-disable-next-line no-param-reassign
        node.data = [
          {
            start: hsp_query_from,
            end: hsp_query_to,
            // franklin $colour-sapphire-blue
            color: '#014371',
            opacity: hsp_identity / 100,
          },
        ];
      }
    },
    [hsp_query_from, hsp_query_to, hsp_identity]
  );

  return (
    <protvista-track
      data-testid="blast-summary-track"
      length={length}
      height={10}
      ref={setTrackData}
      title={`Start: ${hsp_query_from}\nEnd: ${hsp_query_to}\nIdentity: ${hsp_identity}%`}
    />
  );
};

const BlastSummaryHsps: FC<{ hsps: BlastHsp[]; length: number }> = ({
  hsps,
  length,
}) => {
  const [collapsed, setCollapsed] = useState(true);

  const hspsOrderedByScore = hsps.sort(
    (hspA, hspB) => hspB.hsp_bit_score - hspA.hsp_bit_score
  );

  return (
    <div className="data-table__blast-hsp">
      <div className="data-table__blast-hsp__tracks">
        <BlastSummaryTrack hsp={hspsOrderedByScore[0]} length={length} />
        {hspsOrderedByScore.length > 1 &&
          !collapsed &&
          hspsOrderedByScore
            .slice(1)
            .map((hsp) => (
              <BlastSummaryTrack
                hsp={hsp}
                length={length}
                key={`${hsp.hsp_hit_from}-${hsp.hsp_hit_to}`}
              />
            ))}
      </div>
      {hspsOrderedByScore.length > 1 && collapsed && (
        <button type="button" onClick={() => setCollapsed(false)}>{`+${
          hspsOrderedByScore.length - 1
        } more`}</button>
      )}
    </div>
  );
};

const BlastResultTable: FC<{
  data: BlastResults;
  selectedEntries: string[];
  handleSelectedEntries: (rowId: string) => void;
}> = ({ data, selectedEntries, handleSelectedEntries }) => {
  if (!data) {
    return null;
  }
  loadWebComponent('protvista-track', ProtvistaTrack);

  const columns = [
    {
      label: 'Accession',
      name: 'accession',
      render: ({ hit_acc }: BlastHit) => (
        <Link to={`/uniprotkb/${hit_acc}`}>{hit_acc}</Link>
      ),
    },
    {
      label: 'Gene',
      name: 'gene',
      render: ({ hit_uni_gn }: BlastHit) => hit_uni_gn,
    },
    {
      label: 'Organism',
      name: 'organism',
      render: ({ hit_uni_ox, hit_uni_os }: BlastHit) => (
        <Link to={`/taxonomy/${hit_uni_ox}`}>{hit_uni_os}</Link>
      ),
    },
    {
      label: 'Alignment',
      name: 'alignment',
      render: ({ hit_hsps }: BlastHit) => (
        <BlastSummaryHsps hsps={hit_hsps} length={data.query_len} />
      ),
    },
  ];

  return (
    <Fragment>
      <DataTable
        getIdKey={({ hit_acc }: { hit_acc: string }) => hit_acc}
        density={DENSITY_COMPACT}
        columns={columns}
        data={data.hits}
        selectable
        selected={selectedEntries}
        onSelect={handleSelectedEntries}
      />
    </Fragment>
  );
};

export default BlastResultTable;
