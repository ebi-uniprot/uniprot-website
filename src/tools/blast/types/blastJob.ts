import { Status } from './blastStatuses';
import { FormParameters } from './blastFormParameters';

/* Job as defined inside the web application */
interface BaseJob {
  status: Status;
  internalID: string;
  title: string;
  type: 'blast';
  parameters: FormParameters;
  timeCreated: number;
  timeLastUpdate: number;
}

export interface CreatedJob extends BaseJob {
  status: Status.CREATED;
}

export interface FailedJob extends BaseJob {
  status: Status.FAILURE | Status.ERRORED;
  remoteID?: string; // depending on if it failed before or after submission
  timeSubmitted?: number; // depending on if it failed before or after submission
}

export interface RunningJob extends BaseJob {
  status: Status.RUNNING;
  remoteID: string;
  timeSubmitted: number;
}

// Data as BlastResults from './blastResults are not in here all the time to not
// fill the store with too much data
export interface FinishedJob extends BaseJob {
  status: Status.FINISHED;
  remoteID: string;
  timeSubmitted: number;
  timeFinished: number;
  data: {
    hits: number;
  };
}

export type Job = CreatedJob | FailedJob | RunningJob | FinishedJob;
