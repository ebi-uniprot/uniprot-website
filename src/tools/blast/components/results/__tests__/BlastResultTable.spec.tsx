import { fireEvent, screen } from '@testing-library/react';

import customRender from '../../../../../shared/__test-helpers__/customRender';
import { Namespace } from '../../../../../shared/types/namespaces';
import noHits from '../../../../__mocks__/server-jobs/example-empty';
import data from '../../../../__mocks__/server-jobs/example-truncated';
import BlastResultTable from '../BlastResultTable';

describe('BlastResultTable tests', () => {
  it('should render, and toggle the extra HSPs', async () => {
    const { asFragment } = customRender(
      <BlastResultTable
        data={data}
        setSelectedItemFromEvent={jest.fn()}
        setHspDetailPanel={jest.fn()}
        loading={false}
        namespace={Namespace.uniprotkb}
      />
    );
    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByTestId('blast-summary-track')).toBeInTheDocument();
    const toggle = screen.getByRole('button', { name: '+1 more' });
    fireEvent.click(toggle);
    const hspTracks = screen.getAllByTestId('blast-summary-track');
    expect(hspTracks.length).toBe(2);
  });

  it('should render without any hits', () => {
    const { asFragment } = customRender(
      <BlastResultTable
        data={noHits}
        setSelectedItemFromEvent={jest.fn()}
        setHspDetailPanel={jest.fn()}
        loading={false}
        namespace={Namespace.uniprotkb}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByTestId('blast-summary-track')).not.toBeInTheDocument();
  });
});
