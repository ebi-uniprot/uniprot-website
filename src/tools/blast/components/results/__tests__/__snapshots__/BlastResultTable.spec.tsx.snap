// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlastResultTable tests should render the results table 1`] = `
<DocumentFragment>
  <div
    class="data-loader__wrapper"
  >
    <div
      class="data-loader__scroll-container"
    >
      <table
        class="data-table data-table--compact "
      >
        <thead
          class="data-table__header"
        >
          <tr
            class="data-table__row"
          >
            <th
              class="data-table__header-cell"
            >
               
            </th>
            <th
              class="data-table__header-cell "
            >
              Accession
            </th>
            <th
              class="data-table__header-cell "
            >
              Gene
            </th>
            <th
              class="data-table__header-cell "
            >
              Organism
            </th>
            <th
              class="data-table__header-cell "
            >
              Alignment
            </th>
          </tr>
        </thead>
        <tbody
          class="data-table__table__body"
        >
          <tr>
            <td
              class="data-table__cell "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell "
            >
              <a
                href="/uniprotkb/U6MKN0"
              >
                U6MKN0
              </a>
            </td>
            <td
              class="data-table__cell "
            >
              ENH_00046930
            </td>
            <td
              class="data-table__cell "
            >
              <a
                href="/taxonomy/51315"
              >
                Eimeria necatrix
              </a>
            </td>
            <td
              class="data-table__cell "
            >
              <div
                class="data-table__blast-hsp"
              >
                <div
                  class="data-table__blast-hsp__tracks"
                >
                  <protvista-track
                    data-testid="blast-summary-track"
                    height="10"
                    length="30"
                    title="Start: 3
End: 30
Identity: 7.1%"
                  />
                </div>
                <button
                  type="button"
                >
                  +1 more
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
